packages/gatsby-source-contentful/src/normalize.js

```diff
@@ -51,39 +51,46 @@ exports.fixId = fixId
 
 const fixIds = object => _fixIds(object)
 
-// Recursively walk the object model and find any property named `sys`. If it
+// Walk the object model and find any property named `sys`. If it
 // contains an `id` then make sure the id is a string and if it starts with a
 // number, prefix it with (an arbitrarily chosen) `c`, for "contentful".
 // The `front` tracks which objects have been visited to prevent infinite
 // recursion on cyclic structures.
 const _fixIds = (object, front = new Set()) => {
-  if (!object || typeof object !== `object`) {
-    return
-  }
+  const objectsToProcess = [object]
 
-  if (Array.isArray(object)) {
-    object.forEach(v => _fixIds(v, front))
-    return
-  }
+  while (objectsToProcess.length !== 0) {
+    const current = objectsToProcess.pop()
 
-  if (front.has(object)) {
-    return
-  }
+    if (!current || typeof current !== `object`) {
+      continue
+    }
 
-  front.add(object)
-  Object.getOwnPropertyNames(object).forEach(key => {
-    // The `contentful_id` is ours and we want to make sure we don't visit the
-    // same node twice (this is possible if the same node appears in two
-    // separate branches while sharing a common ancestor). This check makes
-    // sure we keep the original `id` preserved in `contentful_id`.
-    if (key === `sys` && !object.sys.contentful_id) {
-      object.sys.contentful_id = object.sys.id
-      object.sys.id = fixId(object.sys.id)
+    if (Array.isArray(current)) {
+      objectsToProcess.push(...current)
+      continue
     }
 
-    _fixIds(object[key], front)
-  })
-  front.delete(object) // Memory vs efficiency
+    if (front.has(current)) {
+      continue
+    }
+
+    front.add(current)
+    Object.keys(current).forEach(key => {
+      // The `contentful_id` is ours and we want to make sure we don't visit the
+      // same node twice (this is possible if the same node appears in two
+      // separate branches while sharing a common ancestor). This check makes
+      // sure we keep the original `id` preserved in `contentful_id`.
+      if (key === `sys` && !current.sys.contentful_id) {
+        current.sys.contentful_id = current.sys.id
+        current.sys.id = fixId(current.sys.id)
+      }
+
+      if (!front.has(current[key])) {
+        objectsToProcess.push(current[key])
```

Comment:
I wonder if we should add all values here, there is a fail-safe at the top of the while loop but this might speed it up a bit if we filter non-objects already.

Reply:
Added check for both falsy values and non-objects.